src/charm.py

```diff
@@ -860,12 +860,19 @@ def _install_snap_packages(self, packages: List[str]) -> None:
         Args:
             packages: list of packages to install.
         """
-        for snap_name, snap_channel in packages:
+        for snap_name, snap_version in packages:
             try:
                 snap_cache = snap.SnapCache()
                 snap_package = snap_cache[snap_name]
 
-                snap_package.ensure(snap.SnapState.Latest, channel=snap_channel)
+                if not snap_package.present:
+                    if snap_version.get("revision"):
+                        snap_package.ensure(
+                            snap.SnapState.Latest, revision=snap_version["revision"]
+                        )
+                        snap_package.hold()
```

Comment:
This will hold to the installed revision indefinitely. We will need to have some kind of update logic in the future.